server/aesdsocket.c

```diff
@@ -257,86 +544,92 @@ int main(int argc, char *argv[])
         
-        /* track buffer index from previous write */
-            recvIndex += recvBytes;
-        /* check for end of packet, \n terminated */
-            if(recvBuf[recvIndex - 1] == '\n')
-                recvDone = true;
+        threadSetup->clientFD   = clientFD;
+        threadSetup->mutex      = &mutex;
+        threadSetup->thread_complete_success = 0;
+        res = pthread_create(&threadSetup->thread_ID, NULL, client_func, (void *)threadSetup);
+        if(res != 0)
+        {
+            /* TODO: pthread create failed to start thread */
+            printf("\nfreeing threadSetup on pthread_create failure\n");
+            free(threadSetup);
+            shutdown(socketFD, SHUT_RDWR);
+            close(socketFD);
+            exit(-1);
         }
 
-        printf("%ld bytes received from %s", recvBytes, clientIP);
-        syslog(LOG_DEBUG, "%ld bytes received from %s", recvBytes, clientIP);
-
-    /* null terminate buffer to write to file as string */
-        buf[totalBytes] = '\0';
-        totalBytes += 1;
+        /* add thread to linked list after creation */
+        SLIST_INSERT_HEAD(&head, threadSetup, entries);
+        printf("\nadded thread to linked list\n");
 
-    /* setup file to write results into */
-        printf("\nfile setup\n");
-        fptr = fopen(OUTPUT_FILE, "a+");
-        printf("write to file\n");
-        fputs(buf, fptr);
-
-    /* clear buf for file readback */
-        memset(buf, 0, MAX_BUF_SIZE * sizeof(buf[0]));
-    /* readback and send contents of file back */
-        fseek(fptr, 0, SEEK_SET);
-        while(fgets(buf, MAX_BUF_SIZE, fptr) != NULL)
+cleanup_threads:
+        /* remove all thread from linked-list, if it is completed */
+        threadSetup = NULL;
+        SLIST_FOREACH_SAFE(threadSetup, &head, entries, temp) 
         {
-            res = send(clientFD, buf, strlen(buf), 0);
-            if(res == -1)
+            if (threadSetup->thread_complete_success) 
             {
-                if(errno == EINTR)
-                    continue;
-                syslog(LOG_ERR, "send failed, see errno for details");
+                pthread_join(threadSetup->thread_ID, NULL);
+                SLIST_REMOVE(&head, threadSetup, thread_data, entries);
+                printf("\nfreeing threadSetup in cleanup_threads\n");
+                free(threadSetup);
+                printf("\tfree complete\n");
             }
         }
+    }
 
-        printf("\ncleanup and close");
-        syslog(LOG_DEBUG, "closing connection from %s", clientIP);
-        fclose(fptr);
-        close(clientFD);
-        printf("\nClosed connection from %s\n", clientIP);
+    printf("\nClean exit interrupt received, closing...\n");
+    syslog(LOG_DEBUG, "closing connection from %s", clientIP);
+    close(socketFD);
+    close(clientFD);
+    remove(OUTPUT_FILE);
+
+    /* cleanup linked list */
+    threadSetup = NULL;
+    while(!SLIST_EMPTY(&head))
+    {
+        threadSetup = SLIST_FIRST(&head);
```

Comment:
@mash9756 you should `pthread_join` here to join any threads which weren't yet complete when the loop exited above.